crates/core/tedge_api/src/entity_store.rs

```diff
@@ -939,114 +940,64 @@ pub enum InitError {
-    // TODO: this is basically manual Deserialize implementation, better impl
-    // Serialize/Deserialize.
-    #[must_use]
-    pub fn new(message: &MqttMessage) -> Option<Self> {
+    pub fn new(message: &MqttMessage) -> Result<Self, serde_json::Error> {
         let topic_id = message
             .topic
             .name
             .strip_prefix(MQTT_ROOT)
-            .and_then(|s| s.strip_prefix('/'))?;
-
-        let payload = parse_entity_register_payload(message.payload_bytes())?;
-
-        let JsonValue::Object(mut properties) = payload else {
-            return None;
-        };
-
-        let Some(JsonValue::String(r#type)) = properties.remove("@type") else {
-            return None;
-        };
-
-        let r#type = match r#type.as_str() {
-            "device" => EntityType::MainDevice,
-            "child-device" => EntityType::ChildDevice,
-            "service" => EntityType::Service,
-            _ => return None,
-        };
-
-        let parent = properties.remove("@parent");
-        let parent = if let Some(parent) = parent {
-            let JsonValue::String(parent) = parent else {
-                return None;
-            };
-            let Ok(parent) = parent.parse() else {
-                return None;
-            };
-            Some(parent)
-        } else {
-            None
-        };
-
-        let parent = if r#type == EntityType::ChildDevice || r#type == EntityType::Service {
-            parent
-        } else {
-            None
-        };
-
-        let health_service = if let Some(value) = properties.remove("@health") {
-            let JsonValue::String(value_str) = value else {
-                return None;
-            };
-            let Ok(health_service) = value_str.parse() else {
-                return None;
-            };
-            Some(health_service)
-        } else {
-            None
-        };
-
-        let entity_id = properties.remove("@id");
-        let entity_id = if let Some(entity_id) = entity_id {
-            let JsonValue::String(entity_id) = entity_id else {
-                return None;
-            };
-            Some(entity_id.into())
-        } else {
-            None
-        };
-
-        let other = properties;
-
-        assert_eq!(other.get("@id"), None);
-        assert_eq!(other.get("@type"), None);
-        assert_eq!(other.get("@parent"), None);
-
-        Some(Self {
-            topic_id: topic_id.parse().ok()?,
-            external_id: entity_id,
-            r#type,
-            parent,
-            health_endpoint: health_service,
-            twin_data: other,
-        })
+            .and_then(|s| s.strip_prefix('/'))
+            .ok_or_else(|| {
+                serde::de::Error::custom("Failed to parse MQTT topic as a valid entity topic")
+            })?;
+        let topic_id = EntityTopicId::from_str(topic_id).map_err(|_| {
+            serde::de::Error::custom("Failed to parse MQTT topic as a valid entity topic")
+        })?;
```

Comment:
I would use `MqttSchema::parse` to extract the topic-id and check this is a metadata channel.

Reply:
Resolved by [8654090](https://github.com/thin-edge/thin-edge.io/pull/3587/commits/8654090f76cf664e56dff2f493927e9a1b324c18). Instead of parsing the `topic-id` here, I've updated the signature in such a way that the already parsed `topic-id` is taken as a parameter, along with the payload.

Reply:
> I've updated the signature in such a way that the already parsed topic-id is taken as a parameter,

Even better.